### PR TITLE
[LLP] Implement the possibility to have a shared StringInterningView [DPP-1042]

### PR DESCRIPTION
**Pull request by tudor-da**
This PR is a preparation for the Participant fusing refactoring and it running a participant in combined mode with a shared StringInterningView.

changelog_begin
changelog_end

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala [hunk 1]
```diff
@@ -94,10 +104,17 @@ class StringInterningView(loadPrefixedEntries: LoadStringInterningEntries)
...
+      val stringsToBeRemoved = idsToBeRemoved.map(raw.idMap)
+
+      raw = raw.copy(
+        map = raw.map.removedAll(stringsToBeRemoved),
+        idMap = raw.idMap.removedAll(idsToBeRemoved),
+      )
+    }
+    else {
       loadPrefixedEntries(raw.lastId, lastStringInterningId)(loggingContext)
         .map(updateView)(ExecutionContext.parasitic)
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala:119]: if we want to have an EC, then it should probably be used here

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala [hunk 2]
```diff
@@ -94,10 +104,17 @@ class StringInterningView(loadPrefixedEntries: LoadStringInterningEntries)
 
   override def update(
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala:105]: as discussed it seems like the time is right for a change of strategy here. Let's make LoadStringInterningEntries as a parameter to update.

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala [hunk 3]
```diff
@@ -94,10 +104,17 @@ class StringInterningView(loadPrefixedEntries: LoadStringInterningEntries)
...
       lastStringInterningId: Int
-  )(implicit loggingContext: LoggingContext): Future[Unit] =
-    if (lastStringInterningId <= raw.lastId) {
-      Future.unit
-    } else {
+  )(implicit loggingContext: LoggingContext, executionContext: ExecutionContext): Future[Unit] =
+    if (lastStringInterningId <= raw.lastId) Future {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala]: Discussed offline and addressed.

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala [hunk 4]
```diff
@@ -94,10 +104,17 @@ class StringInterningView(loadPrefixedEntries: LoadStringInterningEntries)
...
+      val idsToBeRemoved = (lastStringInterningId + 1 to raw.lastId).toSet
+      val stringsToBeRemoved = idsToBeRemoved.map(raw.idMap)
+
+      raw = raw.copy(
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala:109]: Discussed offline. Keeping it as it is.

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala [hunk 5]
```diff
@@ -111,3 +128,29 @@ class StringInterningView(loadPrefixedEntries: LoadStringInterningEntries)
     }
   }
 }
+
+object StringInterningView {
+  def build(
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala]: Fine by me to remove these builders.

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala [hunk 6]
```diff
@@ -28,9 +28,15 @@ trait UpdatingStringInterningView {
...
     * @param lastStringInterningId this is the "version" of the persistent view, which from the StringInterningView can see if it is behind
-    * @return a completion Future: if the view is behind it will load the missing entries from persistence, and update the view state
+    * @return a completion Future:
+    *
+    * * if the view is behind, it will load the missing entries from persistence, and update the view state.
+    *
+    * * if the view is ahead, it will remove all entries with ids greater than the `lastStringInterningId`
     */
-  def update(lastStringInterningId: Int)(implicit loggingContext: LoggingContext): Future[Unit]
+  def update(lastStringInterningId: Int)(
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala:37]: can you please add some comments the update is not thread-safe: not for concurrent usage, and that the internize and read functionality is thread safe.

**Review by nmarton-da — APPROVED** [review]
LGTM, thanks for the changes!
added a comment on making interface description a little bit more explicit about how it shoudl be used

File: ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala (hunk 6 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningView.scala:41]: good point. Thanks for the reminder: added

**Pull request merged**